src/it/scala/io/iohk/ethereum/transactionTest/util/DumpChainActor.scala

```diff
@@ -0,0 +1,174 @@
+package io.iohk.ethereum.transactionTest.util
+
+import java.io.FileWriter
+
+import akka.actor.{Actor, ActorRef, _}
+import akka.util.ByteString
+import io.iohk.ethereum.crypto.kec256
+import io.iohk.ethereum.domain.{BlockHeader, Receipt}
+import io.iohk.ethereum.network.PeerActor.{MessageReceived, SendMessage, Subscribe}
+import io.iohk.ethereum.network.PeerManagerActor.{GetPeers, Peer, Peers}
+import io.iohk.ethereum.network.p2p.messages.PV62._
+import io.iohk.ethereum.network.p2p.messages.PV63.ReceiptImplicits.receiptRlpEncDec
+import io.iohk.ethereum.network.p2p.messages.PV63._
+import io.iohk.ethereum.rlp.RLPImplicitConversions.toRlpList
+import io.iohk.ethereum.rlp.{RLPEncoder, RLPImplicitConversions, encode}
+import org.spongycastle.util.encoders.Hex
+
+import scala.collection.immutable.HashMap
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+class DumpChainActor(peerManager: ActorRef, startBlock: BigInt, maxBlocks: BigInt) extends Actor {
+  var stateNodesHashes: Set[ByteString] = Set.empty
+  var contractNodesHashes: Set[ByteString] = Set.empty
+  var evmCodeHashes: Set[ByteString] = Set.empty
+
+  var blockHeadersStorage: Map[ByteString, BlockHeader] = HashMap.empty
+  var blockHeadersHashes: Seq[ByteString] = Nil
+  var blockBodyStorage: Map[ByteString, BlockBody] = HashMap.empty
+  var blockReceiptsStorage: Map[ByteString, Seq[Receipt]] = HashMap.empty
+  var stateStorage: Map[ByteString, MptNode] = HashMap.empty
+  var contractStorage: Map[ByteString, MptNode] = HashMap.empty
+  var evmCodeStorage: Map[ByteString, ByteString] = HashMap.empty
+
+  var peers: Seq[Peer] = Nil
+
+  override def preStart(): Unit = {
+    context.system.scheduler.scheduleOnce(4 seconds, () => peerManager ! GetPeers)
+  }
+
+  // scalastyle:off
+  override def receive: Receive = {
+    case Peers(p) =>
+      //TODO ask for block headers
+      peers = p.keys.toSeq
+      peers.headOption.foreach { case Peer(_, actor) =>
+        actor ! Subscribe(Set(BlockHeaders.code, BlockBodies.code, Receipts.code, NodeData.code))
+        actor ! SendMessage(GetBlockHeaders(block = Left(startBlock), maxHeaders = maxBlocks, skip = 0, reverse = false))
+      }
+
+    case MessageReceived(m: BlockHeaders) =>
+      val headerHashes = m.headers.map(_.hash)
+      val mptRoots: Seq[ByteString] = m.headers.map(_.stateRoot)
+
+      blockHeadersHashes = m.headers.map(_.hash)
+      m.headers.foreach { h =>
+        blockHeadersStorage = blockHeadersStorage + (h.hash -> h)
+      }
+
+      peers.headOption.foreach { case Peer(_, actor) =>
+        actor ! SendMessage(GetBlockBodies(headerHashes))
+        actor ! SendMessage(GetReceipts(headerHashes.drop(1)))
```

Comment:
Minor detail: Why is `headerHashes.drop(1)` needed? Is it for not asking for the receipts for the genesis block? As `startBlock` is configurable, should we take into account the case where there is no genesis block to drop from `headerHashes`?